File: 백준/Gold/27172. 수 나누기 게임/수 나누기 게임.py
# no. 27172: 수 나누기 게임
# x1, x2, x3, ..., xn
# 만약 xi가 xj로 나누어 떨어지면,
# xi -= 1, xj += 1
# 만약 xj가 xi로 나누어 떨어지면,
# xi += 1, xj -= 1
# 둘다 아니라면, 변화 없음
# 에라토스테네스의 체 응용

from sys import stdin
input = stdin.readline

n = int(input())
nums = {num: 0 for num in map(int, input().split())}

for num in nums.keys():
    tmp = num*2
    while tmp <= 1_000_000:
        if tmp in nums:
            nums[num] += 1
            nums[tmp] -= 1
        tmp += num

print(*nums.values())

File: 백준/Gold/27172. 수 나누기 게임/수 나누기 게임.cc
// no.27172: 수 나누기 게임 (G4)

#include <cstdio>
#include <vector>
using namespace std;

int score[1'000'001];
bool exist[1'000'001];

int main() {
    int n;
    scanf("%d", &n);
    vector<int> nums(n);
    for(int i=0; i<n; i++) {
        scanf("%d", &nums[i]);
        exist[nums[i]] = true;
    }
    for(int num : nums) {
        int tmp = num*2;
        while(tmp <= 1'000'000) {
            if(exist[tmp]) {
                score[tmp]--;
                score[num]++;
            }
            tmp += num;
        }
    }
    for(int num : nums) {
        printf("%d ", score[num]);
    }
    printf("\n");
    return 0;
}